### PR TITLE
[exporter/datadogexporter] add option to submit sum as rate

### PR DESCRIPTION
**Pull request by hardproblems**
**Description:** <Describe what has changed.>
Add an option to datadogexporter to submit sums as rate (to match the behavior of datadog-agent)

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17483

**Testing:**
- added unit tests to metric_exporter_test.go to test submission of sums as rate using different conversion intervals.

**Documentation:**
N/A (datadogexporter README.md links to external site)

@KSerrania @mx-psi @gbbr @knusbaum @amenasria @dineshg13 @aallawala

### CODE REVIEW

File: exporter/datadogexporter/internal/metrics/consumer.go
Comment (aallawala) [inline exporter/datadogexporter/internal/metrics/consumer.go:49]: If this implements translator.Consumer, does that mean the method signature upstream needs to be updated as well to match this?

File: exporter/datadogexporter/config.go
Comment (gbbr) [inline exporter/datadogexporter/config.go:44]:
```suggestion
	// DefaultSumToRateInterval specifies the interval to use when converting a sum to a rate.
```

File: exporter/datadogexporter/config.go
Comment (gbbr) [inline exporter/datadogexporter/config.go:172]: In Go, all documentation must start with the identifier.
```suggestion
	// SubmitAsRate reports whether the sum should be submitted as a rate. The default
	// is false, which means that it will be submitted as a count.
```

File: exporter/datadogexporter/config.go
Comment (gbbr) [inline exporter/datadogexporter/config.go:176]:
```suggestion
	// SumToRateConversionInterval specifies the interval to use when submitting sums
	// as rate. To be used in conjunction with SubmitAsRate.
```

File: exporter/datadogexporter/config.go
Comment (gbbr) [inline exporter/datadogexporter/config.go:521]: Please remove this.

File: exporter/datadogexporter/config.go
Comment (gbbr) [inline exporter/datadogexporter/config.go:519]: Can you remove this empty line please?

**Review by gbbr — CHANGES_REQUESTED** [review]
Just a superficial review to help get started. I will let @mx-psi do the functional review and evaluate whether this change makes sense.

File: exporter/datadogexporter/config.go
Comment (hardproblems) [reply exporter/datadogexporter/config.go]: removed

File: exporter/datadogexporter/config.go
Comment (hardproblems) [reply exporter/datadogexporter/config.go]: removed

File: exporter/datadogexporter/internal/metrics/consumer.go
Comment (gbbr) [inline exporter/datadogexporter/internal/metrics/consumer.go:49]: For me, boolean args are a big no-go. They are ambiguous. When reading the calls it is impossible to understand what they are supposed to mean. Instead, maybe you'll want to use a structure (perhaps called `ConsumerOptions`).

However, I'd recommend to wait for @mx-psi's approval and review before committing any more work to this.

File: exporter/datadogexporter/config.go
Comment (mx-psi) [inline exporter/datadogexporter/config.go:45]: Metrics in OTLP need not be sent at a regular interval, and even if metrics were being sent at a regular interval, this interval may not be a whole number of seconds, which is what our intake currently accepts. Therefore, I am not comfortable accepting this as it is since we are making a transformation that may not make sense semantically.

**Review by mx-psi — CHANGES_REQUESTED** [review]
Hey @hardproblems, thanks for taking the time to open this PR. I would first like to step back and understand why you are interested in this: what behavior are you seeing in-app that you would like to change? What problem are you trying to solve, at a high level?

As you can see on my comment below, introducing this option would make it easy for the Datadog exporter to make transformations that do not make sense semantically, which I don't think it's acceptable.

**Comment by hardproblems** [comment]
>Hey @hardproblems, thanks for taking the time to open this PR. I would first like to step back and understand why you are interested in this: what behavior are you seeing in-app that you would like to change? What problem are you trying to solve, at a high level?

We are using the otel-collector to receive dogstatsd metrics (using the statsdreceiver). Statsdreceiver converts statsd count types to OTLP sums an an aggregation_interval can be configured to aggregate counter value updates in the interval to one datapoint. According to Datadog's [documentation](https://docs.datadoghq.com/metrics/types/?tab=count#submission-types-and-datadog-in-app-types), the datadog-agent will submit dogstatsd `count` to the API as `rate` type. Looking at their [implementation](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/counter.go#L41-L42), it will divide the sum by the interval. We are hoping to mimic this behavior with otel-collector, so that it can replace an internal collector that behaves just like datadog-agent.

**Comment by mx-psi** [comment]
Thanks for the explanation @hardproblems. 

You are right that the Datadog Agent behavior with Dogstatsd counts is to map them to rates, but there are some other key differences. One key difference is that the Datadog Agent performs aggregation of the count at regular intervals, which is what the Agent sets on the interval field. This field does not exist on the OTLP metrics model (in principle, they may not be aggregated on regular intervals). Given the OpenTelemetry Collector uses OTLP as its internal type, in the conversion from Dogstatsd to OTLP performed by the Dogstatsd receiver it will not record the interval (the [`aggregation_interval`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.69.0/receiver/statsdreceiver#configuration) in the statsd receiver terminology). This means that sending the interval is simply not possible so it's not possible to reproduce the exact behavior of the Agent today.

My recommendation is to do this transformation on the webapp instead: you can use the [metric type modifiers](https://docs.datadoghq.com/metrics/custom_metrics/type_modifiers/?tab=count) to convert between rate and count and to use the [rollup](https://docs.datadoghq.com/dashboards/functions/rollup/) function to choose a consistent interval between the two options. I can't assure you that the result will be exactly the same on the two products but this will get you closer to it.

**Comment by hardproblems** [comment]
> Given the OpenTelemetry Collector uses OTLP as its internal type, in the conversion from Dogstatsd to OTLP performed by the Dogstatsd receiver it will not record the interval (the [aggregation_interval](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.69.0/receiver/statsdreceiver#configuration) in the statsd receiver terminology). This means that sending the interval is simply not possible so it's not possible to reproduce the exact behavior of the Agent today.

That's true. However, the aggregation_interval is a static config on the statsdreceiver, and this PR adds the sum_to_rate_conversion_interval as a static config option on the datadogexporter. As long as we take care to configure them the same in the pipeline from statsd->datadog then it behaves the same as the agent.

Re. your recommendation to to the transform on the webapp. All existing graphs / alerts will show a data discontinuity when we flip the existing metric's submission with collector replacement. Communicating to an org of ~2k engineers is non-trivial and not-necessary imho when we add this optional configuration easily.

**Comment by hardproblems** [comment]
@mx-psi i reached out to Datadog chat support and has gotten no updates. What's the best way to proceed here?

**Comment by mx-psi** [comment]
> @mx-psi i reached out to Datadog chat support and has gotten no updates. What's the best way to proceed here?

Please send a follow up message and indicate how urgent this is for you on the message. Some work related to this is being done, but it may take time; if you need something soon we can provide a short-term solution based on metrics attributes.

**Comment by sirianni** [comment]
> Some work related to this is being done, but it may take time; if you need something soon we can provide a short-term solution based on metrics attributes.

@mx-psi @hardproblems - was there any resolution here?  My team is running into the exact same issue when trying to migrate from DD Agent to OTel Collector (or running hybrid setups).

**Comment by mx-psi** [comment]
> > Some work related to this is being done, but it may take time; if you need something soon we can provide a short-term solution based on metrics attributes.
> 
> @mx-psi @hardproblems - was there any resolution here? My team is running into the exact same issue when trying to migrate from DD Agent to OTel Collector (or running hybrid setups).

No resolution unfortunately, if you want to revive this I suggest raising another support ticket from your side so we can internally track that you are also interested in this.

**Comment by RiantZ** [comment]
Hi, recent update: https://github.com/DataDog/datadog-agent/releases/tag/7.80.0

> Add support for the [datadog.metric.as](http://datadog.metric.as/)_type datapoint attribute on OTLP delta sum metrics. When this attribute is set to "rate", the metric is sent to Datadog as a Rate (value divided by interval) instead of a Count. Accepted values are "rate", "count", and "gauge"; unknown values are logged and ignored. This allows users migrating from DogStatsD to OpenTelemetry to preserve rate-type metric behavior.

**Pull request closed without merging**
